[CSM-562] First draft of the proposed V1 API for Updates

adinapoli-iohk opened the pull request:
[Rendered](https://github.com/input-output-hk/cardano-sl/blob/dc51acc8897830ea072b89e543fa7434e8977410/wallet-new/docs/updates.md)

This documents is mostly a brain dump (from me) in the attempt of reworking the current update API for the wallet backend in preparation for the new version of the API (V1).

I have tried to gather the state (and limitations) of the current API by conversations with @domenkozar and @flyingleafe , but I suspect this will be just the beginning of the discussion and not the end.

The TL;DR of the proposed API:

- Updates are divided in optional and mandatory;
- Updates can be skipped, downloaded & applied separately;
- Updates can be downloaded (and checked for data integrity) and installed at a later stage;
- Downloads are now explicit and not implicit.
- Users can check for updates at any time.

As this is now a whiteboard and a chance for all of us as a team to have an active discussion, I would like basically everybody to have a look and share their feedback (especially if what I'm proposing is too naive and doesn't make sense at all 😜 ).

cc @dcoutts @flyingleafe @domenkozar @georgeee @DominikGuzei @darko-mijic

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+-   (1) As a Daedalus user, I want the wallet to prompt me if there is a new update available.
+
+At startup Daedalus would make a call to `GET /api/v1/update` to fetch the last available update. Note
+how this endpoint would **always** return a `WalletSoftwareUpdate`.
```
If there were no updates since Daedalus was first installed, then how?
We might say that "well, there already was updates in history of mainnet anyway" and traverse the blockchain in order to find such old updates, but this seems ugly and I believe it's better to just change how API behaves.

I'm not sure why we really need to display here update which was already applied. Why Daedalus would need this information? Probably in order to display information about this last applied update to the user; but it seems to me that it would be more clear to perform this via separate endpoint (like `GET /api/v1/update/history`), which may return full history of already applied updates.

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+
+-   (1.2) As a Daedalus user, I want to be able to see the download progress for an update.
+
+To accomplish this, one possible idea would be to use [chunked transfer encoding](https://en.wikipedia.org/wiki/Chunked_transfer_encoding)
```
I'm not sure that this is straightforward to implement with Servant. I believe that it's _possible_, but I'm afraid it might require some deep digging into darkest corners of `Servant.Server.Internal` and lots of suffering. I'm happy to be proven wrong, though.

If I'm right, we might do this in dead simple way of providing a separate endpoint for checking download progress. We also might think about something like websocket notifications, but I'm not sure if frontend guys would be happy with such complications.

adinapoli-iohk replied:
Returning chunked http responses it's a pretty standard practice for a web server, but I appreciate might not fall into Servant's standard use case. I'm happy to investigate how painful it is, and I think that if we could use a chunked transfer encoding (unless @domenkozar thinks it's a terrible idea) that would be a win, because we wouldn't need to keep any piece of separate state to check the progress of a download (and thus no separate endpoint) and we could easily treat the download operation as atomic: it's either completed in a single HTTP call, or is not. Which makes keeping the internal state consistent very easy.

flyingleafe replied:
I'm aware of benefits provided by chunked encoding, I just believe that we shouldn't finalize the update api spec until we are sure that we won't have to go to hell and back to implement it.

This won't stop this particular PR to be merged, though. I think that the right way of resolving this is to create separate issue for investigating how chunked encoding can be implemented with Servant, and changing the spec with another PR if it turns out to be really hard.

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+
+This will be ensured by design, as it's only when the user decides to apply the update that the installer is
+moved from the download directory into the folder the Cardano Launcher monitors and uses for the update process.
+
```
Should we include a _separate_ story, when the user _deliberatly_ wants an update to be applied not immediately, but only after manual restart?

adinapoli-iohk replied:
Yes, I think we should as it's a standard case. Conceptually, it doesn't change too much; we will still move the downloaded installer in the directory the launcher will inspect, but we won't trigger a shutdown, as that will be left to the user.

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+                     | Skipped
+                     -- ^ The update has been skipped by the user.
+                     | Proposed
+                     -- ^ The update has been just proposed and is waiting for users' feedback.
```
Shouldn't we add `Downloading` state here? An update which is currently in process of downloading doesn't seem to fit in any of those categories.

adinapoli-iohk replied:
I thought about this, but consider the case where the user is downloading the update but it hard-quits Daedalus. If we store this "transient" state, it will create troubles when we re-access the app, because it would look like the update is downloading, but we really want to start fresh and pretend we never downloaded the upgrade, so I would be tempted to keep `Downloading` out of the sum type and review this in case that's not the right choice.

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+
+If we kept only one single field to store this data, such field might be sensible to data races and *ghost updates*,
+where we could accidentally override data in a way which would cause our internal invariants to be violated.
+
```
It's unclear when the update goes from "latest" slot to "current". Does it happen when user gets this update with `GET /api/v1/update` at the first time? In which cases update in "latest" slot replaces update in "current" slot and in which doesn't? Does skipped updates always get replaced by latest ones? This all needs clarification.

adinapoli-iohk replied:
You are right, people cannot read what's on my mind so I should clarify that 😂

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+        type: object
+      UpdateSeverity:
+        type: enum
+        items: [mandatory, optional]
```
Swagger swears that these lines are incorrect; I believe that you wanted to have
```
type: string
enum: [mandatory, optional]
```
instead.

adinapoli replied:
I will be honest with you, @flyingleafe : I wrote those on top of my head, so of course they are incorrect, and Swagger is right 😅  Well spotted, I will amend it.

adinapoli-iohk replied:
I will be honest with you: I wrote those on top of my head, so of course they are incorrect, and Swagger is right 😅 Well spotted, I will amend it.

adinapoli-iohk commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+-   The wallet access this information and stores the
+    update in a list of updates. The list sorts the updates by
+    oldest first, so adding a new update is a `O(n)` operation where
+    `n` is the number of updates (which is small anyway).
```
Ah! Is there any blocker in moving away from the current scheme but instead monitor the blockchain as we would do normally, but dispatch update blocks to this `ucDownloadedUpdate` variable without waiting for the full download? Am I missing something in the way we push updates to the blockchain? I was under the impression that updates are just normal blocks with special metadata inside which make them identifiable as updates, but this could totally be my distorted way of the world, so please let me know 😜

flyingleafe replied:
@adinapoli-iohk Yes, we can totally populate this variable as soon as an update proposal appears in block, without downloading it; this is just about how things work right now. (And I think we should rename this variable, because it will signal not about downloaded update, but about update available to download).

adinapoli-iohk commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+-   When we want to apply an update, we query `GET /api/update`
+    and this performs a `O(1)` `head` lookup in the list of updates to
+    grab the first (and te oldest), because updates are incremental
+    and needs to be performed in a sequence.
```
Perfect, will amend the docs accordingly as it's important to describe exactly how we are doing things at the moment.

adinapoli-iohk commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,563 @@
+                        | Optional
+
+For hard forks updates we would set the severity to `Mandatory`, for soft forks and other non-critical updates
+we would label an update as `Optional`.
```
Uhm, I see your point, and I don't have a definitive answer.

What I can tell you though, is that starting with just `Optional` sounds like a good compromise, we can always add more items to the enumeration later on _or_ keep `mandatory` in place but not use it.

Note (but you know this very well already, hehe) that our current update system treats all the updates as `Mandatory` anyway, because when we restart Daedalus, the thing get's automatically applied, so that's already an improvement in this area 😀

adinapoli-iohk replied:
Ps. I have found another case for `Mandatory` updates, which would be security-critical updates, for which we definitely want to force an update. What do you think?

adinapoli-iohk commented:
@flyingleafe I have addressed most of your concerns, and the rendered new version is [here](https://github.com/input-output-hk/cardano-sl/blob/83153c47c9217597c03f1e5c7a5596220d04489b/wallet-new/docs/updates.md).

Would you mind having another go and see if it's any better than before? 😉

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,656 @@
+     that Daedalus can warn the user it has a downloaded (but not applied) update.
+
+- When `POST /api/v1/update` is called:
+    - If the input `WalletSoftwareUpdate` doesn't match the _current_ update, ignore the request.
```
We need to return _something_ (error code), it's better to explicitly state that. It would be nice if we decide which error code we should return now.

adinapoli-iohk replied:
It's written it the Swagger spec below, but I do agree I should have reported that here as well, my bad for the inconsistency 😉

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,656 @@
+    - If the input `WalletSoftwareUpdate` does match the _current_ update, carry on as per
+      user story (4) and (4.1), then set the _current_ status to `Applied`.
+
+        - If a _next_ update is available, replace _next_ with _current_.
```
I believe you wanted to say another way round -- replace _current_ with _next_.

adinapoli-iohk replied:
🤦‍♂️

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,656 @@
+        post:
+          description: >
+            Applies this update. It takes as input in the body
+            a JSON representing the `WalletSoftwareUpdate` to perform.
```
This seems like a bit of an overkill. It's sufficient to provide version number or update hash -- either clearly identify an unique update proposal. I believe that even if we go that way, internally we'll compare input update to current one by version number or hash anyway, not as a whole (e. g. to avoid differences in `UpdateState`).

adinapoli-iohk replied:
I see what you mean, and I agree that it feels a bit heavy, but at the same time it's easier for the frontend, because they can simply re-post what they have got from `GET /api/v1/updates`. One middle ground could be to return the computed hash (let's call it `UpdateId`) as part of a `WalletSoftwareUpdate`, so that the frontend chaps can then extract it from the JSON and pass it to this post endpoint, like `POST /api/updates/{updateId}`.

Do you have any strong preference on this one?

flyingleafe replied:
We already have such hash in `WalletSoftwareUpdate`, it's `wsuSha256Checksum`. I don't think we should invent something new.

I see no difference though between using hash or software version, because software version is also a part of `WalletSoftwareUpdate`. Though hash seems to be a little more precise (there shouln't exist two different updates with the same version, but who knows). 

I'm for using either software version or checksum hash for these endpoints instead of the whole datatype. I don't have any strong preference between these two options.

adinapoli-iohk replied:
> We already have such hash in WalletSoftwareUpdate, it's wsuSha256Checksum. I don't think we should invent something new.

@flyingleafe agreed, with the caveat that I have introduced this type as part of the proposal, and there might be extra work (which might not happen for weeks) to include such checksum in an `UpdateProposal` from the core layer, whereas we could go straight to my proposal and maybe still defer such "data integrity" decisions for later. Lame reason or did I convince you? 😀

flyingleafe commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,656 @@
+Such function would gather evidence from an `UpdateProposal` that this is a non-negotiable update.
+
+Note how we don't need to commit just yet whether or not we will treat hard-forks as `Mandatory`, as
+there are good arguments for people wanting to stick with an old version of the chain.
```
So you say that we _will_ stop Daedalus from working if the update is mandatory, but mandatory updates would be only super-severe security ones? Okay, that makes sense, though I think that notion of mandatory update should be discussed more broadly from both philosophical and technical perspectives (e. g. I believe that voting should work differently for super-urgent mandatory updates, and AFAIK we don't have any ideas about it on Core side yet).

adinapoli-iohk replied:
I agree. In principle, this proposal could start with everything marked as `Optional` and we could offload the final decision with the team at a later stage.

adinapoli-iohk commented on wallet-new/docs/updates.md:
```diff
@@ -0,0 +1,656 @@
+        , wsuSeverity        = Optional
+        , wsuUpdateState     = Applied
+        , wsuSha256Checksum  = ...
+        }
```
I think originally, when I wrote the spec, this invariant was helping in the update process in case the user would abruptly hardquit whilst the update process is running, or when a new update arrived. I think, though, since I have separated the current update vs the next update, I think we might be able to go with your suggestion and return a `404`. Let me dwell on it, and I will upgrade the proposal accordingly or share my concerns here :)

adinapoli-iohk commented:
@flyingleafe (Hopefully) final version rendered [here](https://github.com/input-output-hk/cardano-sl/blob/b7fc88058d49ff96c76b0d4efa95363f0d89c24f/wallet-new/docs/updates.md) 😉

flyingleafe commented:
Okay, this looks solid now. I still don't like that you've decided to leave full `WalletSoftwareUpdate`s as `POST/PUT` requests inputs, but I can live with it.
As this is only a documentation, I propose you to clean up all the changes in `default.nix` (cleaning up history with rebase will be the best, but if it turns out to be too cumbersome, a new commit cancelling changes will do), and then we can pretty safely merge this PR without waiting for CI.
(Though we're merging that to your branch, so even this might be unnecessary)

adinapoli commented:
> I still don't like that you've decided to leave full WalletSoftwareUpdates as POST/PUT requests inputs, but I can live with it.

@flyingleafe I left that out with the following rationale:

> the caveat that I have introduced this type [the checksum] as part of the proposal, and there might be extra work (which might not happen for weeks) to include such checksum in an UpdateProposal from the core layer, whereas we could go straight to my proposal and maybe still defer such "data integrity" decisions for later. 

My honest commitment: let's roll out V1, let's have the frontend folks and the core team review the update part, let's amend later if we can with just the updateId. After all up until the Shelley release we can still iterate on V1 up until it won't tick all the satisfaction boxes ;)

adinapoli-iohk commented:
> I still don't like that you've decided to leave full WalletSoftwareUpdates as POST/PUT requests inputs, but I can live with it.

@flyingleafe I left that out with the following rationale:

> the caveat that I have introduced this type [the checksum] as part of the proposal, and there might be extra work (which might not happen for weeks) to include such checksum in an UpdateProposal from the core layer, whereas we could go straight to my proposal and maybe still defer such "data integrity" decisions for later. 

My honest commitment: let's roll out V1, let's have the frontend folks and the core team review the update part, let's amend later if we can with just the updateId. After all up until the Shelley release we can still iterate on V1 up until it won't tick all the satisfaction boxes ;)

adinapoli-iohk commented:
CI is failing on Darwin just due to timeout. Merging this.

georgeee commented:
```
    , wsuSha256Checksum  :: !Sha256Checksum
```

Why SHA-256? Default hash is `Blake2b-256`. And we have it as part of update proposal already (exactly, a hash of installer).


```
data UpdateSeverity = Mandatory
                    | Optional
```

I don't like to distinguish this by soft/hard forks.
If we want to support this field, this should go as a new attribute to update proposal (my opinion).


> I also believe, though, than this restriction was an historical vestige of the fact we were using binary patching (bsdiff) on the executables, but this is no longer true as we ship full installers on S3.

I would argue we should have this option in mind and plan it for post-Shelley. We can't make really frequent updates with 100mb installers, this is absurd.

-----------------

Regarding API in general I dislike us sticking to idea of "latest" update. I would rather have an endpoint to get list of updates (maybe also past updates), and then have all operations like download/etc described in format like `/api/v1/update/<updId>/download`

adinapoli-iohk commented:
Thank you for the excellent review @georgeee , that's really appreciated! Let me try clarifying some of the details in my thought process. I'm very open to changes, and hopefully this will only clarify where I'm coming from 😉 

> Why SHA-256?

No particular reason, I just thought that SHA256 is probably the most common hash function used for checksums and it's easily verifiable, so that's way I was proposing using it. But I don't have any strong feelings, I'm perfectly OK to use `Blake2b-256` as it seems exactly what we need and is a piece of info we have already.

> I don't like to distinguish this by soft/hard forks.
If we want to support this field, this should go as a new attribute to update proposal (my opinion).

I'm open to suggestions. I just thought it was a good idea to have the ability to force an upgrade on the user. There was a similar discussion with @flyingleafe which became a bit philosophical as one may argue that users should be able to choose whether or not they want to support the old or the new chain, so I tend to agree with you distinguishing this via soft/hard forks is not a great idea. The main use case I had in mind, to be honest, was critical security fixes, where definitely we might want to force the user to upgrade. What do you think?

> I would argue we should have this option in mind and plan it for post-Shelley. 

I hear you, and I think using `bsdiff` was a laudable idea. However, when I spoke with @domenkozar , he made very clear this was a lovely idea on paper, but that it cannot work in practice due to the intricacies of supporting multi-platform installers, for example things which writes stuff into Window's registry. Maybe Domen can expand further, but here is my outlook: as long as we are not giving away the possibility of refactoring the _internals_ without changing the _external_ API, we should be good. 

In other terms, we should answer the question: will using this proposed API put ourselves in a future condition of not being able to support binary patching?

> We can't make really frequent updates with 100mb installers, this is absurd.

Maybe 100mb installers are big, this is true, but a very nice advantage over binary patching is that they are _self-contained_. You install one thing, and you are done. Also, this way updates are stateless, which means independent one between the others. If we use binary patching we are effectively introducing a temporal dependency between updates. What would happen if a user install Daedalus 2 years from now? He will have to go throw a painful process of downloading and installing pieces of small incremental updates. What happens if one of the updates along the chain gets corrupted, stale or removed? That would mean users won't be able to update.

Shipping one single installer solves all of this. If we want to prefer stability over size, the fat-installer is the way to go, in my honest opinion.

> Regarding API in general I dislike us sticking to idea of "latest" update. I would rather have an endpoint to get list of updates (maybe also past updates)[..]

I do agree that is not very RESTful to have this weird concept of latest update, and I would be very happy to make the endpoints more homogeneous. However, I guess this is more one for @darko-mijic and @DominikGuzei . If they can work with an endpoint which returns the full list of updates, and use query params filtering to get the latest, I'm _very, very_ happy to make this way more uniform. The question would be, though: do we care?

> [..]and then have all operations like download/etc described in format like /api/v1/update/<updId>/download

Another good idea, and potentially essential if we want in the future to support binary patching. In this case, I think the question should be: Do we really need to download an arbitrary update, even a past one?

dcoutts commented:
Binary diffing is not a high priority. I' reasonably sure that if we ever decide to add it, we could do so transparently to this API since we do the download in the background anyway.

adinapoli-iohk commented:
> Binary diffing is not a high priority. I' reasonably sure that if we ever decide to add it, we could do so transparently to this API since we do the download in the background anyway.

Whilst I agree is not a high priority, in the proposed API we are actually advocating for an explicit download mechanism, so the idea would be that the user would check for new updates in Daedalus and see a download bar progress after he clicked "download". This means conceptually there is a difference between downloading a single, fat installer vs chasing a chain of small incremental updates. I haven't thought too hard about the implications on the endpoints, but I would say that now that we have a versioned API probably the most pragmatic course of action is to get the ball rolling with a decently-sensible update API and the minute we need to do a substantial overhaul, we do so as part of the version `V2` of the API (i.e. mounting the new update API under `/api/v2/...`).

What do you think folks?
